File: src/wp8/Geolocation.cs
/*
Copyright 2015 Intel Corporation

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file 
except in compliance with the License. You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the 
License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, 
either express or implied. See the License for the specific language governing permissions 
and limitations under the License
*/

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Windows.Devices.Geolocation;
using WPCordovaClassLib.Cordova;
using WPCordovaClassLib.Cordova.Commands;
using WPCordovaClassLib.CordovaLib;

namespace Cordova.Extension.Commands
{
    public class IntelXDKGeolocation : BaseCommand
    {
        #region private variables
        Geolocator geolocator;

        private bool polling = false;

        private List<MobiLocationListener> listeners = new List<MobiLocationListener>();

        int successCallbackID = 1;
        int errorCallbackID = 1;
        int maximumAge = 10;
        bool highAccuracy = true;
        int watchId = 0;
        #endregion

        #region constructor
        public IntelXDKGeolocation()
            { }
        #endregion

        #region appMobi.js public methods
        /// <summary>
        /// Get the currrent geolocation position
        /// </summary>
        /// <param name="parameters"></param>
        public async void getCurrentPosition(string parameters)
        {
            string[] args = WPCordovaClassLib.Cordova.JSON.JsonHelper.Deserialize<string[]>(parameters);
            if (args.Length > 3)
            {
                successCallbackID = int.Parse(args[0]);
                errorCallbackID = int.Parse(args[1]);
                maximumAge = int.Parse(args[2]);
                highAccuracy = bool.Parse(args[3]);
            }

            Geoposition geoposition = await GetPosition(maximumAge, highAccuracy);

            Geocoordinate cord = geoposition.Coordinate;

            System.Globalization.CultureInfo ci = new CultureInfo("en-US");

            // AppMobi.Geolocation.prototype.successCB = function (ID, latitude, longitude, altitude, accuracy, altitudeAccuracy, heading, speed, timestamp) {
            var js = string.Format(ci, "javascript:intel.xdk.geolocation.successCB({0},{1},{2},{3},{4},{5},{6},{7},'{8}');", successCallbackID,
                (cord.Latitude != null) ? cord.Latitude : 0.00,
                (cord.Longitude != null) ? cord.Longitude : 0.00,
                (cord.Altitude != null) ? cord.Altitude : 0.00,
                (cord.Accuracy != null) ? cord.Accuracy : 0.00,
                (cord.AltitudeAccuracy != null) ? cord.AltitudeAccuracy : 0.00,
                (cord.Heading != null) ? cord.Heading : 0.00,
                (cord.Speed != null) ? cord.Speed : 0.00,
                cord.Timestamp);

            //DispatchCommandResult(new PluginResult(PluginResult.Status.OK, js));
            InvokeCustomScript(new ScriptCallback("eval", new string[] { js }), true);
            //InjectJS(js);
        }

        public void watchPosition(string parameters)
        {
            string[] args = WPCordovaClassLib.Cordova.JSON.JsonHelper.Deserialize<string[]>(parameters);
            int currentWatchId = watchId;

            successCallbackID = (args.Length > 0) ? int.Parse(args[0]) : 0;
            errorCallbackID = (args.Length > 1) ? int.Parse(args[1]) : 0;
            maximumAge = (args.Length > 3) ? int.Parse(args[3]) : 0;
            highAccuracy = (args.Length > 4) ? bool.Parse(args[4]) : false;

            MobiLocationListener mobiList = new MobiLocationListener(watchId, false, successCallbackID, errorCallbackID, maximumAge, highAccuracy);

            listeners.Add(mobiList);
            watchId++;
            DispatchCommandResult(new PluginResult(PluginResult.Status.OK, currentWatchId));
        }

        public void clearWatch(string parameters)
        {
            string[] args = WPCordovaClassLib.Cordova.JSON.JsonHelper.Deserialize<string[]>(parameters);
            int watchId = 0;
            int.TryParse(args[0], out watchId);

            MobiLocationListener listen = listeners.Where<MobiLocationListener>(x => x.index == watchId).FirstOrDefault();
            listeners.Remove(listen);
            DispatchCommandResult(new PluginResult(PluginResult.Status.OK, watchId));

        }

        public async void pollLocation(string parameters)
        {
            if (!polling)
            {
                polling = true;
                string[] args = WPCordovaClassLib.Cordova.JSON.JsonHelper.Deserialize<string[]>(parameters);
                int watchId = int.Parse(args[0]);

                MobiLocationListener listen = listeners.Where<MobiLocationListener>(x => x.index == watchId).FirstOrDefault();

                if (listen != null)
                {
                    Geoposition geoposition = await GetPosition(listen.maxAge, listen.highAccuracy);

                    Geocoordinate cord = geoposition.Coordinate;

                    System.Globalization.CultureInfo ci = new CultureInfo("en-US");

                    // AppMobi.Geolocation.prototype.successCB = function (ID, latitude, longitude, altitude, accuracy, altitudeAccuracy, heading, speed, timestamp) {
                    var js = string.Format(ci, "javascript:intel.xdk.geolocation.successCB({0},{1},{2},{3},{4},{5},{6},{7},'{8}');", listen.successID,
                        (cord.Latitude != null) ? cord.Latitude : 0.00,
                        (cord.Longitude != null) ? cord.Longitude : 0.00,
                        (cord.Altitude != null) ? cord.Altitude : 0.00,
                        (cord.Accuracy != null) ? cord.Accuracy : 0.00,
                        (cord.AltitudeAccuracy != null) ? cord.AltitudeAccuracy : 0.00,
                        (cord.Heading.HasValue && !Double.IsNaN(cord.Heading.Value)) ? cord.Heading : 0.00,
                        (cord.Speed.HasValue && !Double.IsNaN(cord.Speed.Value)) ? cord.Speed : 0.00,
                        cord.Timestamp);
                    //InjectJS(js);
                    InvokeCustomScript(new ScriptCallback("eval", new string[] { js }), true);
                }
                polling = false;
            }
        }
        #endregion

        #region private methods
        public async Task<Geoposition> GetPosition(int maximumAge, bool highAccuracy)
        {
            Geoposition geoposition=null;

            if (geolocator == null)
                geolocator = new Windows.Devices.Geolocation.Geolocator();

            if (geolocator != null)
            {
                geolocator.DesiredAccuracy = (highAccuracy) ? PositionAccuracy.High : PositionAccuracy.Default;

                //this call will also ask the user for permission to their location
                geoposition = await geolocator.GetGeopositionAsync(
                    maximumAge: TimeSpan.FromSeconds(maximumAge),
                    timeout: TimeSpan.FromSeconds(30)
                );
            }
            return geoposition;
        }
        #endregion

        // Define a listener that responds to location updates
        private class MobiLocationListener
        {
            public int index;
            public int successID, errorID;
            public bool oneTime, locationValid;
            public Geoposition lastLoc;
            public int maxAge;
            public bool highAccuracy;

            public MobiLocationListener(int ind, bool once, int success, int error, int maxAge, bool highAccuracy)
            {
                this.index = ind;
                this.oneTime = once;
                this.successID = success;
                this.errorID = error;
                this.maxAge = maxAge;
                this.highAccuracy = highAccuracy;
            }
            public void onLocationChanged(Geoposition loc)
            {
                // Called when a new location is found by the network location provider.
                //System.out.printf("Loc: %1$f, %2$f, %3$f\n", loc.getLatitude(), loc.getLongitude(), loc.getAltitude());
                lastLoc = loc;
                locationValid = true;
            }
            public String getLast()
            {
                //System.out.println("Geo: getLast, index = " + index);
                if (!locationValid)
                    return "";
                String js = String.Format(
                        "{0},{1},{2},{3},{4},{5},{6},{7},{8},{9}", oneTime ? 1 : 0, successID,
                        lastLoc.Coordinate.Latitude, lastLoc.Coordinate.Longitude, lastLoc.Coordinate.Altitude,
                        lastLoc.Coordinate.Accuracy, 0.00, lastLoc.Coordinate.Heading, lastLoc.Coordinate.Speed,
                        lastLoc.Coordinate.Timestamp);
                //System.out.println("Geo: getLast: js = " + js);
                locationValid = false;
                //if (oneTime)
                //cancelListener(this);
                return js;
            }
            private void fail()
            {
                //injectJS(String.format("javascript:AppMobi.geolocation.errorCB(%1$d);", errorID));
            }
            public void onStatusChanged(String provider, int status)
            {
                if (status == 0)
                {   // Out of service.
                    fail();
                }
                else if (status == 1)
                {
                    // TEMPORARILY_UNAVAILABLE.
                }
                else
                {
                    // Available
                }
            }
            public void onProviderEnabled(String provider) { }
            public void onProviderDisabled(String provider)
            {
                fail();
            }
        }
    }
}

File: www/geolocation.js
/*
Copyright 2015 Intel Corporation

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file 
except in compliance with the License. You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the 
License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, 
either express or implied. See the License for the specific language governing permissions 
and limitations under the License
*/

var exec = require('cordova/exec'),
	Coords = require('./coords'),
	Position = require('./position');

		/**
		 * This class contains acceleration information
		 * @constructor
		 * @param {Number} x The force applied by the device in the x-axis.
		 * @param {Number} y The force applied by the device in the y-axis.
		 * @param {Number} z The force applied by the device in the z-axis.
		 * @param {boolean} doRotate If true, rotate axes based on device rotation.
		 */

	module.exports = {
		watchIDs: new Array(),
		successCBs: new Array(),
		errorCBs: new Array(),
		pollID: -1,

		getCurrentPosition: function(successCallback, errorCallback, options) {
			if( (successCallback == undefined || successCallback==''))
			{
				throw(new Error("Error: getCurrentPosition has the following required parameters: successCallback."));
			}

			var successID = this.getSuccessId(successCallback);
			var errorID = errorCallback == undefined ? -1 : this.getErrorId(errorCallback);
			var enableHighAccuracy = true;
			var maximumAge = -1;
			
			if (options != undefined) {
				if (options.maximumAge != undefined)
					maximumAge = options.maximumAge;
				if (options.enableHighAccuracy != undefined)
					enableHighAccuracy = options.enableHighAccuracy;
			}
			
			//AppMobiGeolocation.printMessage("getCurrentPosition: " + id + ", pollID = " + this.pollID);
			//var id = AppMobiGeolocation.getCurrentPosition(successID, errorID, maximumAge, enableHighAccuracy);
			exec(function(id) {
				intel.xdk.geolocation.watchIDs.push(id);
			}, null, "IntelXDKGeolocation", "getCurrentPosition", [successID, errorID, maximumAge, enableHighAccuracy]);
			
			if (this.pollID == -1) {
				this.pollID = setInterval("intel.xdk.geolocation.poll()", 200);	// Poll every 200 msecs.
			}
		},
	
		poll: function() {
			var i, to = 0, len = this.watchIDs.length;

			for (i = 0; i < len; ++i) {
				exec(function(loc) {
					if (loc != "") {	
						loc = new String(loc);	// Convert to actual string.
						var vals = loc.split(",");
						var once = vals[0];
						intel.xdk.geolocation.successCB(vals[1],vals[2],vals[3],vals[4],vals[5],vals[6],vals[7],vals[8], vals[9]);
						if (once == 0)
							intel.xdk.geolocation.watchIDs[to++] = intel.xdk.geolocation.watchIDs[i];
						else if (once == 1)
							intel.xdk.geolocation.clearWatch(intel.xdk.geolocation.watchIDs[i]);
						
					} else
						intel.xdk.geolocation.watchIDs[to++] = intel.xdk.geolocation.watchIDs[i];				

				}, null, "IntelXDKGeolocation", "pollLocation", [this.watchIDs[i]]);
			}

			//if (to < len)
			//	this.watchIDs.splice(to);
		},
			
		getSuccessId: function(successCallback) {
			var i, len = this.successCBs.length;
			for (i = 0; i < len; ++i) {
				if (this.successCBs[i] == successCallback)
					return i;
			}
			len = this.successCBs.push(successCallback);
			return len - 1;
		},
			
		getErrorId: function(errorCallback) {
			var i, len = this.errorCBs.length;
			for (i = 0; i < len; ++i) {
				if (this.errorCBs[i] == errorCallback)
					return i;
			}
			len = this.errorCBs.push(errorCallback);
			return len - 1;
		},
		
		successCB: function(ID, latitude, longitude, altitude, accuracy, altitudeAccuracy, heading, speed, timestamp) {
			var fun = this.successCBs[ID];
			var coords = new intel.xdk.coords.Coords(latitude, longitude, altitude, accuracy, altitudeAccuracy, heading, speed);
			var p = new intel.xdk.position.Position(coords, timestamp);
			fun(p);
		},
		
		errorCB: function(ID) {
			var fun = this.errorCBs[ID];
			fun();
		},
		
		watchPosition: function(successCallback, errorCallback, options) {
			if( (successCallback == undefined || successCallback==''))
			{
				throw(new Error("Error: watchPosition has the following required parameters: successCallback."));
			}
			
			var successID = this.getSuccessId(successCallback);
			var errorID = errorCallback == undefined ? -1 : this.getErrorId(errorCallback);
			var freq = 10000;
			var enableHighAccuracy = true;
			var maximumAge = -1;
			var returnId;

			if (options != undefined) {
				if (options.timeout != undefined)
					freq = options.timeout;
				if (options.maximumAge != undefined)
					maximumAge = options.maximumAge;
				if (options.enableHighAccuracy != undefined)
				enableHighAccuracy = options.enableHighAccuracy;
			}
			
			//var id = AppMobiGeolocation.watchPosition(successID, errorID, freq, maximumAge, enableHighAccuracy);
			
			exec(function(id) {
				intel.xdk.geolocation.watchIDs.push(id);
				returnId = id;
			}, null, "IntelXDKGeolocation", "watchPosition", [successID, errorID, freq, maximumAge, enableHighAccuracy]);

			//AppMobiGeolocation.printMessage("watchPosition: " + id + ", pollID = " + this.pollID);
			//this.watchIDs.push(id);
			if (this.pollID == -1) {
				//AppMobiGeolocation.printMessage("About to schedule poll()");
				this.pollID = setInterval("intel.xdk.geolocation.poll()", 200);	// Poll every 200 msecs.
			}
			return returnId;
		},
			
		clearWatch: function(id) {
			//AppMobiGeolocation.clearWatch(id);

				exec(function(id) {
					var index = intel.xdk.geolocation.watchIDs.indexOf(id);
					if (index != -1) {
						intel.xdk.geolocation.watchIDs.splice(index, 1);
						if (intel.xdk.geolocation.watchIDs.length == 0 && id != -1) {
							clearInterval(id);
							intel.xdk.geolocation.pollID = -1;
						}
					}
				}, null, "IntelXDKGeolocation", "clearWatch", [id]);
		}
}